cmd/bootstrap/run/epochs_test.go

```diff
@@ -0,0 +1,102 @@
+		require.NoError(testifyT, err)
+		partnerNodes, err := common.ReadFullPartnerNodeInfos(log, partnerWeights, partnerDir)
+		require.NoError(testifyT, err)
+
+		allNodeIds := make(flow.IdentityList, 0)
+		for _, node := range append(internalNodes, partnerNodes...) {
+			allNodeIds = append(allNodeIds, node.Identity())
+		}
+
+		rootSnapshot := unittest.RootSnapshotFixture(allNodeIds)
+		allIdentities, err := rootSnapshot.Identities(filter.Any)
+		require.NoError(testifyT, err)
+
+		rapid.Check(testifyT, func(t *rapid.T) {
+			numberOfNodesToInclude := rapid.IntRange(0, 3).Draw(t, "nodes-to-include")
+			numberOfNodesToExclude := rapid.UintRange(0, 3).Draw(t, "nodes-to-exclude")
+
+			// we specifically omit collection nodes from the exclusion list since we have a specific
+			// check that there must be a valid cluster of collection nodes.
+			excludedNodes, err := allIdentities.Filter(
+				filter.Not(filter.HasRole[flow.Identity](flow.RoleCollection))).Sample(numberOfNodesToExclude)
+			require.NoError(t, err)
+			excludeNodeIds := excludedNodes.NodeIDs()
+			// an eligible participant is a current epoch participant with a weight greater than zero that has not been explicitly excluded
+			eligibleEpochIdentities := allIdentities.Filter(filter.And(
+				filter.IsValidCurrentEpochParticipant,
+				filter.HasWeightGreaterThanZero[flow.Identity],
+				filter.Not(filter.HasNodeID[flow.Identity](excludeNodeIds...))))
+
+			expectedNodeIds := make(map[cadence.String]struct{})
+			includeNodeIds := unittest.IdentifierListFixture(numberOfNodesToInclude)
+			for _, nodeID := range eligibleEpochIdentities.NodeIDs().Union(includeNodeIds) {
+				expectedNodeIds[cadence.String(nodeID.String())] = struct{}{}
+			}
+
+			epochProtocolState, err := rootSnapshot.EpochProtocolState()
+			require.NoError(t, err)
+			currentEpochCommit := epochProtocolState.EpochCommit()
+			expectedDKGIndexMap := make(map[cadence.String]cadence.Int)
+			for nodeID, index := range currentEpochCommit.DKGIndexMap {
+				expectedDKGIndexMap[cadence.String(nodeID.String())] = cadence.NewInt(index)
+			}
+
+			args, err := GenerateRecoverTxArgsWithDKG(
+				log,
+				internalNodes,
+				2, // number of collection clusters
+				currentEpochCommit.Counter+1,
+				flow.Localnet,
+				100,   // staking function length, in views
+				4000,  // recovery epoch length, in views
+				60*60, // recovery epoch duration, in seconds
+				false, // unsafe overwrite
+				currentEpochCommit.DKGIndexMap,
+				currentEpochCommit.DKGParticipantKeys,
+				currentEpochCommit.DKGGroupKey,
+				excludeNodeIds,
+				includeNodeIds,
+				rootSnapshot,
+			)
+			require.NoError(t, err)
+
+			// dkg index map
+			for _, pair := range args[10].(cadence.Dictionary).Pairs {
+				expectedIndex, ok := expectedDKGIndexMap[pair.Key.(cadence.String)]
+				require.True(t, ok)
+				require.Equal(t, expectedIndex, pair.Value.(cadence.Int))
+			}
+			// node ids
+			for _, nodeId := range args[11].(cadence.Array).Values {
```

Comment:
```suggestion
			nodeIDsArgValues := args[11].(cadence.Array).Values
			assert.Equal(t, len(nodeIDsArgValues), len(expectedNodeIds))
			for _, nodeId := range  {
```
similar suggestion here